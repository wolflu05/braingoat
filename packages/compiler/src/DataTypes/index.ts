import { Int } from "./Int";

export const DataTypes = {
  int: Int,
};
