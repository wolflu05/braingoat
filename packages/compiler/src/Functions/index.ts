import { print } from "./print";

export const Functions = {
  print,
};
