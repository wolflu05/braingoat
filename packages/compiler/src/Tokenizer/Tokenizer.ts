import { generateSplitRegex } from "../utils";
import { OPERATOR_MAP } from "../AST";

export type LineType = {
  line: number;
  col: number;
};

export type TokenType = {
  value: string;
  source: LineType;
};

export class Tokenizer {
  static tokenize(code: string[]) {
    const tokens: TokenType[] = [];
    const splitters = [" ", "(", ")", "{", "}", "[", "]", ",", "<", ">", "/*", "*/", ...Object.keys(OPERATOR_MAP), "="];

    for (const [lineNum, line] of Object.entries(code)) {
      const re = generateSplitRegex(splitters);
      let m, p;
      while (((p = re.lastIndex), (m = re.exec(line)))) {
        // This is necessary to avoid infinite loops with zero-width matches
        if (m.index === re.lastIndex) {
          re.lastIndex++;
        }

        // token between match
        tokens.push({
          value: line.substring(p, m.index),
          source: { line: +lineNum, col: p },
        });

        // actual matched token
        tokens.push({
          value: m[0],
          source: { line: +lineNum, col: m.index },
        });
      }

      // last token in line
      tokens.push({
        value: line.substring(p),
        source: { line: +lineNum, col: p },
      });
    }

    return tokens.filter((token) => !["", " ", "\n"].includes(token.value));
  }
}
