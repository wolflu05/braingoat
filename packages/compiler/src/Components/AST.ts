import { Braingoat, ErrorType } from "../Braingoat";
import { findIndexAt, isValidVariableName } from "../utils";
import { LineType, TokenType } from "./Tokenizer";

export enum TOKEN_TYPES {
  DECLARATION = "DECLARATION",
  FUNCTION_CALL = "FUNCTION_CALL",
  EXPRESSION = "EXPRESSION",
  VALUE_LITERAL = "VALUE_LITERAL",
  ASSIGNMENT = "ASSIGNMENT",
  VARIABLE_LITERAL = "VARIABLE_LITERAL",
}

export type DECLARATION_OPTIONS = {
  variableName: string;
  type: string;
  value: AST;
};

export type FUNCTION_CALL_OPTIONS = {
  functionName: string;
  parameters: TokenType[];
  block: AST[];
};

export enum EXPRESSION_OPERATOR_TYPE {
  ADD = "ADD",
  SUBTRACT = "SUBTRACT",
  MULTIPLY = "MULTIPLY",
  DIVIDE = "DIVIDE",
  POWER = "POWER",
  EQ = "EQ",
  NEQ = "NEQ",
}
export const OPERATOR_MAP = {
  "+": EXPRESSION_OPERATOR_TYPE.ADD,
  "-": EXPRESSION_OPERATOR_TYPE.SUBTRACT,
  "*": EXPRESSION_OPERATOR_TYPE.MULTIPLY,
  "/": EXPRESSION_OPERATOR_TYPE.DIVIDE,
  "^": EXPRESSION_OPERATOR_TYPE.POWER,
  "==": EXPRESSION_OPERATOR_TYPE.EQ,
  "!=": EXPRESSION_OPERATOR_TYPE.NEQ,
};
export type EXPRESSION_OPTIONS = {
  op: EXPRESSION_OPERATOR_TYPE;
  leftNode: AST;
  rightNode: AST;
};

export type VALUE_LITERAL_OPTIONS =
  | {
      type: "INT";
      value: number;
    }
  | {
      type: "INTLIST";
      value: number[];
    };
export type VARIABLE_LITERAL_OPTIONS = {
  name: string;
};

export type ASSIGNMENT_OPTIONS = {
  variableName: string;
  value: AST;
};

export type TOKEN_OPTIONS =
  | DECLARATION_OPTIONS
  | FUNCTION_CALL_OPTIONS
  | EXPRESSION_OPTIONS
  | VALUE_LITERAL_OPTIONS
  | VARIABLE_LITERAL_OPTIONS
  | ASSIGNMENT_OPTIONS;

export class AST {
  type: TOKEN_TYPES;
  tokenOptions: TOKEN_OPTIONS;
  source: LineType;

  constructor(type: TOKEN_TYPES, tokenOptions: TOKEN_OPTIONS, source: LineType) {
    this.type = type;
    this.tokenOptions = tokenOptions;
    this.source = source;
  }

  static parseExpression(tokens: TokenType[], braingoat: Braingoat, startIndex = 0): [AST, number] | null {
    const token = tokens[startIndex];
    if (!token) {
      return null;
    }

    if (token.value === "(") {
      const exprIndex = [];
      const stack = [];
      let k = startIndex;

      for (; k < tokens.length; k++) {
        if (stack.length === 1) {
          exprIndex.push(k);
        }

        if (tokens[k].value === "(") {
          stack.push("(");
        } else if (tokens[k].value === ")") {
          stack.pop();
        }

        if (stack.length === 0) break;
      }

      if (stack.length !== 0) {
        braingoat.throwError(ErrorType.SyntaxError, "Expected )", tokens[k - 1]);
      }

      if (exprIndex.length !== 4) {
        braingoat.throwError(
          ErrorType.SyntaxError,
          "Invalid length expression, expressions can only have two operants wrapped in brackets",
          tokens[k - 1],
        );
      }

      const op = OPERATOR_MAP[tokens[exprIndex[1]].value as keyof typeof OPERATOR_MAP];
      const leftNode = this.parseExpression(tokens, braingoat, exprIndex[0]);
      const rightNode = this.parseExpression(tokens, braingoat, exprIndex[2]);

      if (!op) {
        braingoat.throwError(
          ErrorType.SyntaxError,
          `Invalid operator ${tokens[exprIndex[1]].value}`,
          tokens[exprIndex[1]],
        );
      }
      if (!leftNode) {
        braingoat.throwError(ErrorType.SyntaxError, "Expected left-side operant", tokens[exprIndex[1] - 1]);
      }
      if (!rightNode) {
        braingoat.throwError(ErrorType.SyntaxError, "Expected right-side operant", tokens[exprIndex[1]]);
      }

      return [
        new AST(
          TOKEN_TYPES.EXPRESSION,
          {
            op,
            leftNode: leftNode?.[0] || null,
            rightNode: rightNode?.[0] || null,
          },
          tokens[exprIndex[1]].source,
        ),
        k + 1,
      ];
    }

    if (!Number.isNaN(+token.value)) {
      return [new AST(TOKEN_TYPES.VALUE_LITERAL, { type: "INT", value: +token.value }, token.source), startIndex + 1];
    }

    if (isValidVariableName(token.value)) {
      return [new AST(TOKEN_TYPES.VARIABLE_LITERAL, { name: token.value }, token.source), startIndex + 1];
    }

    return null;
  }

  static parse(tokens: TokenType[], braingoat: Braingoat) {
    const tree: AST[] = [];

    for (let i = 0; i < tokens.length; ) {
      let nextIndex = i;

      // DECLARATION
      if (tokens[i + 2]?.value === "=") {
        const expression = this.parseExpression(tokens, braingoat, i + 3);
        if (!expression) {
          braingoat.throwError(ErrorType.SyntaxError, "Invalid expression", tokens[i + 3]);
        }

        tree.push(
          new AST(
            TOKEN_TYPES.DECLARATION,
            {
              variableName: tokens[i + 1].value,
              type: tokens[i].value,
              value: expression[0],
            },
            tokens[i].source,
          ),
        );
        nextIndex = expression[1];
      }

      // ASSIGNMENT
      else if (tokens[i + 1]?.value === "=") {
        const expression = this.parseExpression(tokens, braingoat, i + 2);
        if (!expression) {
          braingoat.throwError(ErrorType.SyntaxError, "Invalid expression", tokens[i + 2]);
        }

        tree.push(
          new AST(
            TOKEN_TYPES.ASSIGNMENT,
            {
              variableName: tokens[i].value,
              value: expression[0],
            },
            tokens[i].source,
          ),
        );
        nextIndex = expression[1];
      }

      // FUNCTION_CALL
      else if (tokens[i + 1]?.value === "(") {
        const endTokenIndex = findIndexAt(i + 1, tokens, (t) => t.value === ")");
        if (endTokenIndex === -1) {
          braingoat.throwError(ErrorType.SyntaxError, "Expected )", tokens[i + 1]);
        }

        const params = tokens.slice(i + 2, endTokenIndex);
        const wrongIdx = params.findIndex((v, i) => ((i + 1) % 2 === 0 ? v.value !== "," : false));

        if (wrongIdx !== -1) {
          braingoat.throwError(ErrorType.SyntaxError, `Expected ,`, params[wrongIdx]);
        }

        // check for block function
        let block: TokenType[] = [];
        if (tokens[endTokenIndex + 1]?.value === "{" && tokens.length > endTokenIndex + 2) {
          const stack = [];
          let k = endTokenIndex + 1;

          for (; k < tokens.length; k++) {
            if (tokens[k].value === "{") {
              stack.push("{");
            } else if (tokens[k].value === "}") {
              stack.pop();
            }

            if (stack.length === 0) break;
            block.push(tokens[k]);
          }
          if (stack.length !== 0) {
            braingoat.throwError(ErrorType.SyntaxError, "Expected }", tokens[k - 1]);
          }

          nextIndex = k + 1;
        } else {
          nextIndex = endTokenIndex + 1;
        }

        tree.push(
          new AST(
            TOKEN_TYPES.FUNCTION_CALL,
            {
              functionName: tokens[i].value,
              parameters: params.filter((x) => x.value !== ","),
              block: AST.parse(block.slice(1), braingoat),
            },
            tokens[i].source,
          ),
        );
      }

      // UNPROCESSED KEYWORD
      else {
        braingoat.throwError(ErrorType.SyntaxError, "Unprocessed keyword", tokens[i]);
      }

      if (i === nextIndex) {
        nextIndex++;
      }
      i = nextIndex;
    }

    return tree;
  }
}
