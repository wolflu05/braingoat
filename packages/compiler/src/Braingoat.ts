import { AST } from "./AST";
import { Emitter } from "./Emitter";
import { LineType, Tokenizer, TokenType } from "./Tokenizer";
import { CompileError, SyntaxError } from "./utils/errors";

export enum ErrorType {
  SyntaxError = "SyntaxError",
  CompileError = "CompileError",
}

export class Braingoat {
  code: string[];
  bfCode: string;

  constructor(code: string) {
    this.code = code.split("\n");
    this.bfCode = "";
  }

  compile() {
    const tokens = Tokenizer.tokenize(this.code);
    const ast = AST.parse(tokens, this);
    const emitter = new Emitter(ast, this).emit();
    this.bfCode = emitter.code;

    return this;
  }

  throwError(type: ErrorType, message: string, token?: TokenType | LineType): never {
    if (!token) token = { startLine: 0, startCol: 0, endLine: 0, endCol: 0 };
    let source;
    if ("source" in token) {
      source = token.source as LineType;
    } else {
      source = token;
    }

    let stack = "";
    stack += `${this.code[source.startLine - 2] || ""}\n`;
    stack += `${this.code[source.startLine - 1] || ""}\n`;
    stack += `${" ".repeat(source.startCol)}${"^".repeat(source.endCol - source.startCol)} L${source.startLine}:${
      source.startCol
    }-${source.endCol}\n`;
    stack += `${" ".repeat(source.startCol)}${type}: ${message}`;

    if (type == ErrorType.SyntaxError) {
      const error = new SyntaxError(message, stack, source);
      throw error;
    }

    if (type == ErrorType.CompileError) {
      const error = new CompileError(message, stack, source);
      throw error;
    }

    throw new Error(stack);
  }
}
