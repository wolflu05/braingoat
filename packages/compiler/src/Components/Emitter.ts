import { Braingoat, ErrorType } from "../Braingoat";
import { DataTypes } from "../DataTypes";
import { Int } from "../DataTypes/Int";
import { Functions } from "../Functions";
import { BasicDataType, memoryPositionType } from "../Types/BasicDataType";
import { COMMANDS, VALID_CHARS } from "../utils/commands";
import { AST, DECLARATION_OPTIONS, FUNCTION_CALL_OPTIONS, TOKEN_TYPES } from "./AST";
import { LineType, TokenType } from "./Tokenizer";

export class Emitter {
  memoryAllocation: BasicDataType[];
  currentMemoryAddress: number;
  code: string;
  ast: AST[];
  braingoat: Braingoat;

  constructor(ast: AST[], braingoat: Braingoat) {
    this.memoryAllocation = [];
    this.currentMemoryAddress = 0;
    this.code = "";
    this.ast = ast;
    this.braingoat = braingoat;
  }

  emit() {
    for (const node of this.ast) {
      // VARIABLE DECLARATION
      if (node.type === TOKEN_TYPES.DECLARATION) {
        const { type, variableName, value } = node.tokenOptions as DECLARATION_OPTIONS;
        this.registerVariable(type, variableName, value, node.source);
      }

      // FUNCTION CALL
      else if (node.type === TOKEN_TYPES.FUNCTION_CALL) {
        const { functionName, parameters, block } = node.tokenOptions as FUNCTION_CALL_OPTIONS;
        this.callFunction(functionName, parameters, block, node.source);
      }
    }

    return this;
  }

  movePointerTo(address: number, addToCode = true) {
    let res = "";

    while (this.currentMemoryAddress !== address) {
      if (this.currentMemoryAddress < address) {
        res += COMMANDS.MOVE_POINTER_RIGHT;
        this.currentMemoryAddress++;
      } else {
        res += COMMANDS.MOVE_POINTER_LEFT;
        this.currentMemoryAddress--;
      }
    }

    if (addToCode) this.addCode(res);

    return res;
  }

  getNextNEmpty(length: number): memoryPositionType {
    let pos = this.memoryAllocation
      .sort((a, b) => a.position[0] - b.position[0])
      .findIndex((_, i, l) => {
        if (!l[i + 1]) return true;
        return l[i + 1].position[0] - l[i].position[1] - 1 >= length;
      });

    if (pos === -1) {
      pos = 0;
    } else {
      pos = this.memoryAllocation[pos].position[1] + 1;
    }

    return [pos, pos + length - 1];
  }

  registerVariable(type: string, name: string, value: string, source: LineType) {
    if (!(type in DataTypes)) {
      this.braingoat.throwError(ErrorType.CompileError, `${type} is no valid data type`, source);
    }

    this.memoryAllocation.push(new DataTypes[type as keyof typeof DataTypes](this, name, value, source));
  }

  getVariable(name: string | BasicDataType, source: LineType): BasicDataType | never {
    if (name instanceof BasicDataType) {
      return name;
    }

    const variable = this.memoryAllocation.find((x) => x.name === name);

    if (variable === undefined) {
      this.braingoat.throwError(ErrorType.CompileError, `Variable ${name} is not defined`, source);
    }

    return variable;
  }

  callFunction(name: string, args: TokenType[], block: AST[], source: LineType) {
    if (!(name in Functions)) {
      this.braingoat.throwError(ErrorType.CompileError, `${name} is no valid function`, source);
    }

    Functions[name as keyof typeof Functions](this, args, block, source);
  }

  codeBuilder(code: TemplateStringsArray, ...variables: ReadonlyArray<string | number | Int | Function>) {
    code.forEach((c, i) => {
      this.addCode(c);

      if (variables[i]) {
        if (typeof variables[i] === "string") {
          this.addCode(variables[i] as string);
        } else if (typeof variables[i] === "number") {
          this.addCode(this.movePointerTo(variables[i] as number, false));
        } else if (variables[i] instanceof Int) {
          this.addCode(this.movePointerTo((variables[i] as Int).position[0], false));
        } else if (typeof variables[i] === "function") {
          this.addCode((variables[i] as Function)());
        }
      }
    });
  }

  addCode(...code: string[]) {
    this.code += code
      .join("")
      .split("")
      .filter((x) => VALID_CHARS.includes(x))
      .join("");
  }
}
