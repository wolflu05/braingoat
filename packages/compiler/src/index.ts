import { Braingoat } from "./Braingoat";
try {
  const bg = new Braingoat(`
  Int a = 1
  Int b = 1
  
  IntList<3> k = [10,20,30]
  k[(a + b)] = 42

  print(k[(a+(b-1))])
  `).compile();
  // k[1] = 5
  // b = (a + b)
  // Int a = (1 + (i * j)  )
  console.log(bg.bfCode);
} catch (err) {
  console.log((err as Error).stack);
}
