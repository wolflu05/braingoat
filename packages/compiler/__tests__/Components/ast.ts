import { Braingoat } from "../../src/Braingoat";
import { AST } from "../../src/Components/AST";
import { TokenType } from "../../src/Components/Tokenizer";

const astCases: Array<[string, TokenType[], string, boolean]> = [
  ["empty token list", [], "", true],
  [
    "tokenize two variables",
    [
      { value: "abc", source: { line: 0, col: 0 } },
      { value: "i", source: { line: 0, col: 4 } },
      { value: "=", source: { line: 0, col: 6 } },
      { value: "0", source: { line: 0, col: 8 } },
      { value: "int", source: { line: 1, col: 0 } },
      { value: "j", source: { line: 1, col: 4 } },
      { value: "=", source: { line: 1, col: 6 } },
      { value: "1", source: { line: 1, col: 8 } },
    ],
    "abc i = 0\nint j = 1",
    true,
  ],
  [
    "should parse functions",
    [
      { value: "abc", source: { line: 0, col: 0 } },
      { value: "(", source: { line: 0, col: 3 } },
      { value: ")", source: { line: 0, col: 4 } },
    ],
    "abc()",
    true,
  ],
  [
    "should parse functions with parameter",
    [
      { value: "abc", source: { line: 0, col: 0 } },
      { value: "(", source: { line: 0, col: 3 } },
      { value: "1", source: { line: 0, col: 4 } },
      { value: ")", source: { line: 0, col: 5 } },
    ],
    "abc(1)",
    true,
  ],
  [
    "should parse functions with parameters",
    [
      { value: "abc", source: { line: 0, col: 0 } },
      { value: "(", source: { line: 0, col: 3 } },
      { value: "x1", source: { line: 0, col: 4 } },
      { value: ",", source: { line: 0, col: 6 } },
      { value: "1", source: { line: 0, col: 7 } },
      { value: ",", source: { line: 0, col: 8 } },
      { value: "2", source: { line: 0, col: 9 } },
      { value: ",", source: { line: 0, col: 10 } },
      { value: "3", source: { line: 0, col: 11 } },
      { value: ")", source: { line: 0, col: 12 } },
    ],
    "abc(x1,1,2,3)",
    true,
  ],
  [
    "should parse functions with block",
    [
      { value: "abc", source: { line: 0, col: 0 } },
      { value: "(", source: { line: 0, col: 3 } },
      { value: ")", source: { line: 0, col: 4 } },
      { value: "{", source: { line: 0, col: 6 } },
      { value: "}", source: { line: 1, col: 0 } },
    ],
    "abc() {\n}",
    true,
  ],
  [
    "should parse functions with parameter and block",
    [
      { value: "abc", source: { line: 0, col: 0 } },
      { value: "(", source: { line: 0, col: 3 } },
      { value: "1", source: { line: 0, col: 4 } },
      { value: ")", source: { line: 0, col: 5 } },
      { value: "{", source: { line: 0, col: 7 } },
      { value: "}", source: { line: 2, col: 0 } },
    ],
    "abc(1) {\n\n}",
    true,
  ],
  [
    "should parse functions with parameters and block",
    [
      { value: "abc", source: { line: 0, col: 0 } },
      { value: "(", source: { line: 0, col: 3 } },
      { value: "x1", source: { line: 0, col: 4 } },
      { value: ",", source: { line: 0, col: 6 } },
      { value: "1", source: { line: 0, col: 7 } },
      { value: ",", source: { line: 0, col: 8 } },
      { value: "2", source: { line: 0, col: 9 } },
      { value: ",", source: { line: 0, col: 10 } },
      { value: "3", source: { line: 0, col: 11 } },
      { value: ")", source: { line: 0, col: 12 } },
      { value: "{", source: { line: 0, col: 14 } },
      { value: "abc", source: { line: 1, col: 2 } },
      { value: "(", source: { line: 1, col: 4 } },
      { value: ")", source: { line: 1, col: 5 } },
      { value: "{", source: { line: 1, col: 0 } },
      { value: "}", source: { line: 1, col: 6 } },
      { value: "}", source: { line: 2, col: 0 } },
    ],
    "abc(x1,1,2,3) {\n  abc(){}\n}",
    true,
  ],
  [
    "should parse functions with parameters and nested blocks",
    [
      { value: "abc", source: { line: 0, col: 0 } },
      { value: "(", source: { line: 0, col: 3 } },
      { value: "x1", source: { line: 0, col: 4 } },
      { value: ",", source: { line: 0, col: 6 } },
      { value: "1", source: { line: 0, col: 7 } },
      { value: ",", source: { line: 0, col: 8 } },
      { value: "2", source: { line: 0, col: 9 } },
      { value: ",", source: { line: 0, col: 10 } },
      { value: "3", source: { line: 0, col: 11 } },
      { value: ")", source: { line: 0, col: 12 } },
      { value: "{", source: { line: 0, col: 14 } },
      { value: "abc", source: { line: 1, col: 7 } },
      { value: "(", source: { line: 1, col: 10 } },
      { value: ")", source: { line: 1, col: 11 } },
      { value: "{", source: { line: 1, col: 13 } },
      { value: "def", source: { line: 2, col: 9 } },
      { value: "(", source: { line: 2, col: 12 } },
      { value: "1", source: { line: 2, col: 13 } },
      { value: ",", source: { line: 2, col: 14 } },
      { value: "2", source: { line: 2, col: 15 } },
      { value: ")", source: { line: 2, col: 16 } },
      { value: "{", source: { line: 2, col: 18 } },
      { value: "abc", source: { line: 3, col: 11 } },
      { value: "(", source: { line: 3, col: 14 } },
      { value: ")", source: { line: 3, col: 15 } },
      { value: "}", source: { line: 4, col: 9 } },
      { value: "}", source: { line: 5, col: 7 } },
      { value: "}", source: { line: 6, col: 5 } },
    ],
    `abc(x1,1,2,3) {
      abc() {
        def(1,2) {
          abc()
        }
      }
    }`,
    true,
  ],
  [
    "Function call without ending )",
    [
      { value: "abc", source: { line: 0, col: 0 } },
      { value: "(", source: { line: 0, col: 3 } },
      { value: "x1", source: { line: 0, col: 4 } },
      { value: ",", source: { line: 0, col: 6 } },
      { value: "1", source: { line: 0, col: 7 } },
      { value: ",", source: { line: 0, col: 8 } },
      { value: "2", source: { line: 0, col: 9 } },
      { value: ",", source: { line: 0, col: 10 } },
      { value: "3", source: { line: 0, col: 11 } },
    ],
    "abc(x1,1,2,3",
    false,
  ],
  [
    "Function call without ending block }",
    [
      { value: "abc", source: { line: 0, col: 0 } },
      { value: "(", source: { line: 0, col: 3 } },
      { value: "1", source: { line: 0, col: 4 } },
      { value: ",", source: { line: 0, col: 5 } },
      { value: "2", source: { line: 0, col: 6 } },
      { value: ",", source: { line: 0, col: 7 } },
      { value: "3", source: { line: 0, col: 8 } },
      { value: ")", source: { line: 0, col: 9 } },
      { value: "{", source: { line: 0, col: 11 } },
    ],
    "abc(1,2,3) {",
    false,
  ],
  [
    "Function call with more {",
    [
      { value: "abc", source: { line: 0, col: 0 } },
      { value: "(", source: { line: 0, col: 3 } },
      { value: "1", source: { line: 0, col: 4 } },
      { value: ",", source: { line: 0, col: 5 } },
      { value: "2", source: { line: 0, col: 6 } },
      { value: ",", source: { line: 0, col: 7 } },
      { value: "3", source: { line: 0, col: 8 } },
      { value: ")", source: { line: 0, col: 9 } },
      { value: "{", source: { line: 0, col: 11 } },
      { value: "{", source: { line: 0, col: 12 } },
      { value: "}", source: { line: 0, col: 13 } },
    ],
    "abc(1,2,3) {{}",
    false,
  ],
  ["Unprocessed keyword", [{ value: "{", source: { line: 0, col: 0 } }], "{", false],
  ["Unprocessed word keyword", [{ value: "hello", source: { line: 0, col: 0 } }], "hello", false],
];

describe("tokenize", () => {
  for (const [name, tokens, code, shouldBeValid] of astCases) {
    test(name, () => {
      const braingoat = new Braingoat(code);

      if (shouldBeValid) {
        const ast = AST.parse(tokens, braingoat);
        expect(ast).toMatchSnapshot();
      } else {
        const testReachFn = jest.fn();

        try {
          AST.parse(tokens, braingoat);
          testReachFn();
        } catch (error) {
          expect((error as Error).stack).toMatchSnapshot();
        }
        expect(testReachFn).not.toHaveBeenCalled();
      }
    });
  }
});
