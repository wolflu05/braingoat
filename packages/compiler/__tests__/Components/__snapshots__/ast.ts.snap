// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenize Function call with more { 1`] = `
"
abc(1,2,3) {{}
             ^ L0:14
             SyntaxError: Expected }"
`;

exports[`tokenize Function call without ending ) 1`] = `
"
abc(x1,1,2,3
^ L0:1
SyntaxError: Expected tokens )"
`;

exports[`tokenize Function call without ending block } 1`] = `
"
abc(1,2,3) {
           ^ L0:12
           SyntaxError: Unprocessed keyword"
`;

exports[`tokenize Unprocessed keyword 1`] = `
"
{
^ L0:1
SyntaxError: Unprocessed keyword"
`;

exports[`tokenize Unprocessed word keyword 1`] = `
"
hello
^ L0:1
SyntaxError: Unprocessed keyword"
`;

exports[`tokenize empty token list 1`] = `Array []`;

exports[`tokenize should parse functions 1`] = `
Array [
  AST {
    "source": Object {
      "col": 0,
      "line": 0,
    },
    "tokenOptions": Object {
      "block": Array [],
      "functionName": "abc",
      "parameters": Array [],
    },
    "type": "FUNCTION_CALL",
  },
]
`;

exports[`tokenize should parse functions with block 1`] = `
Array [
  AST {
    "source": Object {
      "col": 0,
      "line": 0,
    },
    "tokenOptions": Object {
      "block": Array [],
      "functionName": "abc",
      "parameters": Array [],
    },
    "type": "FUNCTION_CALL",
  },
]
`;

exports[`tokenize should parse functions with parameter 1`] = `
Array [
  AST {
    "source": Object {
      "col": 0,
      "line": 0,
    },
    "tokenOptions": Object {
      "block": Array [],
      "functionName": "abc",
      "parameters": Array [
        AST {
          "source": Object {
            "col": 4,
            "line": 0,
          },
          "tokenOptions": Object {
            "type": "INT",
            "value": 1,
          },
          "type": "VALUE_LITERAL",
        },
      ],
    },
    "type": "FUNCTION_CALL",
  },
]
`;

exports[`tokenize should parse functions with parameter and block 1`] = `
Array [
  AST {
    "source": Object {
      "col": 0,
      "line": 0,
    },
    "tokenOptions": Object {
      "block": Array [],
      "functionName": "abc",
      "parameters": Array [
        AST {
          "source": Object {
            "col": 4,
            "line": 0,
          },
          "tokenOptions": Object {
            "type": "INT",
            "value": 1,
          },
          "type": "VALUE_LITERAL",
        },
      ],
    },
    "type": "FUNCTION_CALL",
  },
]
`;

exports[`tokenize should parse functions with parameters 1`] = `
Array [
  AST {
    "source": Object {
      "col": 0,
      "line": 0,
    },
    "tokenOptions": Object {
      "block": Array [],
      "functionName": "abc",
      "parameters": Array [
        AST {
          "source": Object {
            "col": 4,
            "line": 0,
          },
          "tokenOptions": Object {
            "index": null,
            "name": "x1",
          },
          "type": "VARIABLE_LITERAL",
        },
        AST {
          "source": Object {
            "col": 7,
            "line": 0,
          },
          "tokenOptions": Object {
            "type": "INT",
            "value": 1,
          },
          "type": "VALUE_LITERAL",
        },
        AST {
          "source": Object {
            "col": 9,
            "line": 0,
          },
          "tokenOptions": Object {
            "type": "INT",
            "value": 2,
          },
          "type": "VALUE_LITERAL",
        },
        AST {
          "source": Object {
            "col": 11,
            "line": 0,
          },
          "tokenOptions": Object {
            "type": "INT",
            "value": 3,
          },
          "type": "VALUE_LITERAL",
        },
      ],
    },
    "type": "FUNCTION_CALL",
  },
]
`;

exports[`tokenize should parse functions with parameters and block 1`] = `
Array [
  AST {
    "source": Object {
      "col": 0,
      "line": 0,
    },
    "tokenOptions": Object {
      "block": Array [
        AST {
          "source": Object {
            "col": 2,
            "line": 1,
          },
          "tokenOptions": Object {
            "block": Array [],
            "functionName": "abc",
            "parameters": Array [],
          },
          "type": "FUNCTION_CALL",
        },
      ],
      "functionName": "abc",
      "parameters": Array [
        AST {
          "source": Object {
            "col": 4,
            "line": 0,
          },
          "tokenOptions": Object {
            "index": null,
            "name": "x1",
          },
          "type": "VARIABLE_LITERAL",
        },
        AST {
          "source": Object {
            "col": 7,
            "line": 0,
          },
          "tokenOptions": Object {
            "type": "INT",
            "value": 1,
          },
          "type": "VALUE_LITERAL",
        },
        AST {
          "source": Object {
            "col": 9,
            "line": 0,
          },
          "tokenOptions": Object {
            "type": "INT",
            "value": 2,
          },
          "type": "VALUE_LITERAL",
        },
        AST {
          "source": Object {
            "col": 11,
            "line": 0,
          },
          "tokenOptions": Object {
            "type": "INT",
            "value": 3,
          },
          "type": "VALUE_LITERAL",
        },
      ],
    },
    "type": "FUNCTION_CALL",
  },
]
`;

exports[`tokenize should parse functions with parameters and nested blocks 1`] = `
Array [
  AST {
    "source": Object {
      "col": 0,
      "line": 0,
    },
    "tokenOptions": Object {
      "block": Array [
        AST {
          "source": Object {
            "col": 7,
            "line": 1,
          },
          "tokenOptions": Object {
            "block": Array [
              AST {
                "source": Object {
                  "col": 9,
                  "line": 2,
                },
                "tokenOptions": Object {
                  "block": Array [
                    AST {
                      "source": Object {
                        "col": 11,
                        "line": 3,
                      },
                      "tokenOptions": Object {
                        "block": Array [],
                        "functionName": "abc",
                        "parameters": Array [],
                      },
                      "type": "FUNCTION_CALL",
                    },
                  ],
                  "functionName": "def",
                  "parameters": Array [
                    AST {
                      "source": Object {
                        "col": 13,
                        "line": 2,
                      },
                      "tokenOptions": Object {
                        "type": "INT",
                        "value": 1,
                      },
                      "type": "VALUE_LITERAL",
                    },
                    AST {
                      "source": Object {
                        "col": 15,
                        "line": 2,
                      },
                      "tokenOptions": Object {
                        "type": "INT",
                        "value": 2,
                      },
                      "type": "VALUE_LITERAL",
                    },
                  ],
                },
                "type": "FUNCTION_CALL",
              },
            ],
            "functionName": "abc",
            "parameters": Array [],
          },
          "type": "FUNCTION_CALL",
        },
      ],
      "functionName": "abc",
      "parameters": Array [
        AST {
          "source": Object {
            "col": 4,
            "line": 0,
          },
          "tokenOptions": Object {
            "index": null,
            "name": "x1",
          },
          "type": "VARIABLE_LITERAL",
        },
        AST {
          "source": Object {
            "col": 7,
            "line": 0,
          },
          "tokenOptions": Object {
            "type": "INT",
            "value": 1,
          },
          "type": "VALUE_LITERAL",
        },
        AST {
          "source": Object {
            "col": 9,
            "line": 0,
          },
          "tokenOptions": Object {
            "type": "INT",
            "value": 2,
          },
          "type": "VALUE_LITERAL",
        },
        AST {
          "source": Object {
            "col": 11,
            "line": 0,
          },
          "tokenOptions": Object {
            "type": "INT",
            "value": 3,
          },
          "type": "VALUE_LITERAL",
        },
      ],
    },
    "type": "FUNCTION_CALL",
  },
]
`;

exports[`tokenize tokenize two variables 1`] = `
Array [
  AST {
    "source": Object {
      "col": 0,
      "line": 0,
    },
    "tokenOptions": Object {
      "type": "abc",
      "value": AST {
        "source": Object {
          "col": 8,
          "line": 0,
        },
        "tokenOptions": Object {
          "type": "INT",
          "value": 0,
        },
        "type": "VALUE_LITERAL",
      },
      "variableName": "i",
      "variableOptions": null,
    },
    "type": "DECLARATION",
  },
  AST {
    "source": Object {
      "col": 0,
      "line": 1,
    },
    "tokenOptions": Object {
      "type": "int",
      "value": AST {
        "source": Object {
          "col": 8,
          "line": 1,
        },
        "tokenOptions": Object {
          "type": "INT",
          "value": 1,
        },
        "type": "VALUE_LITERAL",
      },
      "variableName": "j",
      "variableOptions": null,
    },
    "type": "DECLARATION",
  },
]
`;
